pkg/arvo/app/graph-store.hoon

```diff
@@ -0,0 +1,544 @@
+  |^
+  ?>  (team:title our.bowl src.bowl)
+  =^  cards  state
+    ?+  mark           (on-poke:def mark vase)
+        %graph-update  (graph-update !<(update:store vase))
+    ==
+  [cards this]
+  ::
+  ++  graph-update
+    |=  =update:store
+    ^-  (quip card _state)
+    |^
+    ?>  ?=(%0 -.update)
+    ?-  -.q.update
+        %add-graph          (add-graph +.q.update)
+        %remove-graph       (remove-graph +.q.update)
+        %add-nodes          (add-nodes p.update +.q.update)
+        %remove-nodes       (remove-nodes p.update +.q.update)
+        %add-signatures     (add-signatures p.update +.q.update)
+        %remove-signatures  (remove-signatures p.update +.q.update)
+        %add-tag            (add-tag +.q.update)
+        %remove-tag         (remove-tag +.q.update)
+        %archive-graph      (archive-graph +.q.update)
+        %unarchive-graph    (unarchive-graph +.q.update)
+        %run-updates        (run-updates +.q.update)
+        %keys               ~|('cannot send %keys as poke' !!)
+        %tags               ~|('cannot send %tags as poke' !!)
+        %tag-queries        ~|('cannot send %tag-queries as poke' !!)
+    ==
+    ::
+    ++  add-graph
+      |=  [=resource:store =graph:store mark=(unit mark:store)]
+      ^-  (quip card _state)
+      ?<  (~(has by archive) resource)
+      ?<  (~(has by graphs) resource)
+      ?>  (validate-graph graph mark)
+      :_  %_  state
+              graphs       (~(put by graphs) resource [graph mark])
+              update-logs  (~(put by update-logs) resource (gas:orm-log ~ ~))
+          ==
+      %-  zing
+      :~  (give [/updates /keys ~] [%add-graph resource graph mark])
+          ?~  mark  ~
+          =/  wire  (weld /graph (en-path:res resource))
+          =/  =rave:clay  [%sing %b [%da now.bowl] /[u.mark]]
+          [%pass wire %arvo %c %warp our.bowl [%home `rave]]~
+      ==
+    ::
+    ++  remove-graph
+      |=  =resource:store
+      ^-  (quip card _state)
+      ?<  (~(has by archive) resource)
+      ?>  (~(has by graphs) resource)
+      :-  (give [/updates /keys ~] [%remove-graph resource])
+      %_  state
+          graphs       (~(del by graphs) resource)
+          update-logs  (~(del by update-logs) resource)
+      ==
+    ::
+    ++  add-nodes
+      |=  $:  =time
+              =resource:store
+              nodes=(map index:store node:store)
+          ==
+      ^-  (quip card _state)
+      |^
+      =/  [=graph:store mark=(unit mark:store)]
+        (~(got by graphs) resource)
+      =/  =update-log:store  (~(got by update-logs) resource)
+      =.  update-log
+        (put:orm-log update-log time [%0 time [%add-nodes resource nodes]])
+      ::
+      :-  (give [/updates]~ [%add-nodes resource nodes])
+      %_  state
+          update-logs  (~(put by update-logs) resource update-log)
+          graphs
+        %+  ~(put by graphs)
+          resource
+        [(add-node-list resource graph mark ~(tap by nodes)) mark]
```

Comment:
These indexes probably need to be sorted.  Otherwise, behavior could differ based on mug order in the `nodes` map.  As an example, if someone tried to add a post and then a comment on that post, presumably if the order were reversed, validation would fail.

Reply:
Fixed this in 65eb5678